### ISSUE TITLE
Implement assertNotFired, assertFired and assertErrbackFired on SynchronousTestCase

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6105]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6105|
|Type|enhancement|
|Created|2012-10-12 20:25:35Z|
|Branch|https://github.com/twisted/twisted/tree/sync-deferred-testing-helper-6105-2|

In many tests it is useful to deal with Deferreds in a synchronous manner. The current idiom is:

```
result = []
d.addCallback(result.append)
self.assertTrue(result)
```

This is icky (a technical term!). We should instead have assertions on `twisted.trial.unittest.SynchronousTestCase` that can run against Deferreds without relying on the reactor (i.e. no need to return a Deferred from the test).

Attachments:

* [stuff.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fc6/fc63f8aada9986ad684d09fae71554032119ea14/461199c1ad8f7065fc9e6d0c290b4bacf93b6daf.py) (1764 bytes) - added by mithrandi on 2012-11-17 19:24:07Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6105 6105
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_sync_deferred_testing_helper_6105_2 branches/sync-deferred-testing-helper-6105-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1350073535000000 1350073535000000
changetime__1354634963000000 1354634963000000
version__None None
owner__exarkun exarkun
cc__mithrandi
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

Additionally the trial howto should be fixed: currently it tells you to return a Deferred even for a test that doesn't use the reactor. Using these APIs in the documentation would be better.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_2"></a>|@mithrandi commented|
|-|-|

Attaching some stuff I hacked up, although it's untested, and exarkun says there's already a few implementations of this in Twisted.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [36454]) Branching to 'sync-deferred-testing-helper-6105'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [36455]) SynchronousTestCase.resultOf

refs [#6105](https://github.com/twisted/twisted/issues/6105)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [36457]) Switch to resultOf

refs [#6105](https://github.com/twisted/twisted/issues/6105)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Once I get back to this, I think it's going to be `successResultOf`, `failureResultOf`, and `assertNotFired`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [36477]) Branching to 'sync-deferred-testing-helper-6105-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/sync-deferred-testing-helper-6105-2)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @exarkun|
|-|-|

1. Pyflakes:
twisted/web/test/test_agent.py:18: 'reactor' imported but unused
twisted/web/test/test_agent.py:18: 'interfaces' imported but unused
twisted/web/test/test_agent.py:593: redefinition of unused 'error' from line 15

2.
```
+        if len(result) == 0:
...
+        if len(result) == 1:
```
It's a stylistic point but I generally prefer to do 'if not result' 'if result' in those cases where you know it's a list. I guess our coding standard doesn't make everything explicit. Your call, etc.

3. I don't think you can put a L{} in @raise, epytext seems to be complaining about it. There are also lots of invalid ref in test_assertions.

4.
```
+<h2>Testing Deferreds Without The Reactor</h2>
```

No capitalization? :)

5. The howto file can't be parsed properly, there is missing `</em>`

6. The 2 elif clauses in successResultOf and failureResultOf don't seem to be tested.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

Thanks!

> Pyflakes ...

Fixed in r36487 and r36491

> It's a stylistic point but I generally prefer to do 'if not result' 'if result' in those cases where you know it's a list

Hummm.  Okay.  Another stylistic point, I generally prefer avoiding 'if not x' when I can somehow write 'if x'.  I don't see a good way to rewrite (2 out of 3 of) these as 'if result' instead of 'if not result'.  However, 'if not result' doesn't seem too bad in those two cases.  Fixed in r36493.

> I don't think you can put a L{} in @raise

I always mess that one up.  Fixed in r36488.

> No capitalization? :)

Fixed in r36489.

> The howto file can't be parsed properly

Fixed in r36490.

> The 2 elif clauses in successResultOf and failureResultOf don't seem to be tested

Embarrassing.  Fixed in r36492.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set owner to @exarkun|
|-|-|

7.
```
+    failure  = Failure(Exception("Bad times"))
```

There is an extra space before "=".

8. There are many new invalid ref in epytext. At least for the public methods:
```
twisted.trial._synctest._Assertions.successResultOf:594 invalid ref to Deferred
twisted.trial._synctest._Assertions.successResultOf:594 invalid ref to Deferred.callback
twisted.trial._synctest._Assertions.successResultOf:594 invalid ref to Deferred.errback
twisted.trial._synctest._Assertions.successResultOf:594 invalid ref to Deferred
twisted.trial._synctest._Assertions.successResultOf:594 invalid ref to Deferred
twisted.trial._synctest._Assertions.failureResultOf:624 invalid ref to Deferred
twisted.trial._synctest._Assertions.failureResultOf:624 invalid ref to Deferred.callback
twisted.trial._synctest._Assertions.failureResultOf:624 invalid ref to Deferred.errback
twisted.trial._synctest._Assertions.failureResultOf:624 invalid ref to Deferred
twisted.trial._synctest._Assertions.failureResultOf:624 invalid ref to Deferred
twisted.trial._synctest._Assertions.assertNoResult:657 invalid ref to Deferred
twisted.trial._synctest._Assertions.assertNoResult:657 invalid ref to Deferred.callback
twisted.trial._synctest._Assertions.assertNoResult:657 invalid ref to Deferred.errback
twisted.trial._synctest._Assertions.assertNoResult:657 invalid ref to Deferred
twisted.trial._synctest._Assertions.assertNoResult:657 invalid ref to Deferred
twisted.trial._synctest._Assertions.assertNoResult:657 invalid ref to Deferred
twisted.trial._synctest._Assertions.assertNoResult:657 invalid ref to Deferred
```

Please merge once fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [36502]) Fix epytext link errors

refs [#6105](https://github.com/twisted/twisted/issues/6105)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [36503]) remove extra whitespace


refs [#6105](https://github.com/twisted/twisted/issues/6105)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [36506]) Merge sync-deferred-testing-helper-6105-2

Author: exarkun
Reviewer: therve
Fixes: [#6105](https://github.com/twisted/twisted/issues/6105)


Add `successResultOf`, `failureResultOf`, and `assertNoResult` to `SynchronousTestCase`, to help
developers write tests for `Deferred`-using code.
